src/systems/audio-debug-system.js

```diff
@@ -0,0 +1,323 @@
+    return vec4( rgb, circle);
+  }
+
+  vec4 att(float d, vec4 circle, float maxDistance, float refDistance, float rolloffFactor, int distanceModel) {
+    // Calculate attenuation
+    float attenuation = 1.0;
+    if (distanceModel == LINEAR) {
+      attenuation = att_linear(d, rolloffFactor, refDistance, maxDistance);
+    } else if (distanceModel == INVERSE) {
+      attenuation = att_inverse(d, rolloffFactor, refDistance);
+    } else if (distanceModel == EXPONENTIAL) {
+      attenuation = att_exponential(d, rolloffFactor, refDistance);
+    }
+    attenuation = clamp(attenuation, 0.0, 1.0);
+
+    // Waves
+    float v = sin((d * 2.0 * kPi) - (time * 0.005)) + 1.0;
+    float waves = circle.a * attenuation * v * vNormal.y;
+    
+    // Output final color.
+    return vec4( circle.rgb, waves );
+  }
+
+  void main() {
+    // Draw background
+    vec4 background = vec4(1.0, 1.0, 1.0, 0.0);
+
+    for (int i=0; i<count; i++) {
+      vec2 center = sourcePosition[i].xz - vUv;
+
+      // Calculate distance to (0,0).
+      float d = length( center );
+
+      // Optional start Offset.
+      vec2 orientation = normalize(sourceOrientation[i].xz);
+      float ang = atan(-orientation.x, orientation.y) * kInvPi * 0.5;
+      // Rotate to start drawing facing front
+      ang -= 0.5;
+      float startOffset = mod(ang, 1.0);
+
+      // Draw inner cone
+      float innerAngle = coneInnerAngle[i] * kDegToRad * kInvPi * 0.5;
+      float innerStartAngle = startOffset + innerAngle * 0.5;
+      vec4 innerLayer = circle(center, d, innerAngle, 10000.0, 1.0, colorInner, innerStartAngle);
+      innerLayer = att(d, innerLayer, maxDistance[i], refDistance[i], rolloffFactor[i], distanceModel[i]);
+      background = mix(background, innerLayer, innerLayer.a);
+
+      // Draw outer cone
+      float outerAngle = coneOuterAngle[i] * kDegToRad * kInvPi * 0.5;
+      float outerAngleDiffHalf = (outerAngle - innerAngle) * 0.5;
+      vec4 outerLayer1 = circle(center, d, outerAngleDiffHalf, 10000.0, 1.0, colorOuter, innerStartAngle + outerAngleDiffHalf);
+      outerLayer1 = att(d, outerLayer1, maxDistance[i], refDistance[i], rolloffFactor[i], distanceModel[i]);
+      background = mix(background, outerLayer1, outerLayer1.a);
+      vec4 outerLayer2 = circle(center, d, outerAngleDiffHalf, 10000.0, 1.0, colorOuter, innerStartAngle - innerAngle);
+      outerLayer2 = att(d, outerLayer2, maxDistance[i], refDistance[i], rolloffFactor[i], distanceModel[i]);
+      background = mix(background, outerLayer2, outerLayer2.a * (clipped[i] ? 0.0 : 1.0));
+
+      // Draw base
+      vec4 baseLayer = circle(center, d, 1.0, 1.0, 0.1, vec3(0.5, 0.5, 0.5), 0.0);
+      background = mix(background, baseLayer, baseLayer.a);
+
+      // Draw gain
+      float g = clamp(gain[i], 0.0, 1.0);
+      vec4 gainLayer = circle(center, d, g, 1.0, 0.5, colorGain, startOffset);
+      background = mix(background, gainLayer, gainLayer.a);
+      if (gain[i] > 1.0) {
```

Comment:
In case its helpful, I was originally going to say that from what I read using `fwidth` inside a conditional is sort of undefined behavior, because the threads that are inactive in the conditional are still sampled but that register would have an arbitrary value. But in this case all threads are going to be executing the conditional in the same way since its only dependent on the uniforms, which are by definition uniform across all threads.

Reply:
I didn't know that, I've been reading and it makes sense the way modern GPUs work but yeah being uniform dependant should be ok. I'm not sure how optimized this shader is, I guess it can be pushed a little bit more, not sure if the time/perfomance gain makes sense though.

Reply:
Yeah this is only for debugging anyway so I don't think we need to really worry about perf on it very much at all. Just hadn't heard of `fwidth` before and noticed this while reading about it.